### ISSUE TITLE
File permissions are not synced

### ISSUE THREAD

**Issue by C0rn3j** [nextcloud/server#6725]
Am using Arch Linux for clients, Arch Linux ARM for server(s) and NC 12.0.3.

### Steps to reproduce
1. Create a file on machine1
2. Change its permissions - for example run chmod +x on it
3. Let NC sync it
4. On machine2 let NC sync, and check the file permissions.

### Expected behaviour
File has same permissions on both machines. (not sure if owner:group should also be synced though)

### Actual behaviour
```
Machine one: $ ls -lah ~/Nextcloud/company/Scripty/Install/upload.sh
-rwxr-xr-x 1 c0rn3j c0rn3j 143 Sep 25 16:23 ~/Nextcloud/company/Scripty/Install/upload.sh

Machine two: $ ls -lah ~/Nextcloud/company/Scripty/Install/upload.sh 
-rw-r--r-- 1 c0rn3j c0rn3j 143 Sep 25 16:23 ~/Nextcloud/company/Scripty/Install/upload.sh
```

---

This breaks my scripts.

**Comment by nickvergessen** [comment]
Yeah this is not possible, otherwise the files would be executable on the server as well, which in itself is a security thread

**Comment by mimaoffice** [comment]
Is it not possible so "solve" this problem by saving the files permissions into the DB?

**Comment by rio-** [comment]
As suggested by @mimaoffice the files in Nextclouds data directory don't have to have the permissions. Furthermore, I don't even see a security risk of having `+x` set for files owned by Nextcloud, you need to already have access to a shell on the server to be able to execute it and `+x` doesn't escalate your privileges like `+s` could.

Can this issue be reopened?

**Comment by gengwg** [comment]
Having same issue. I wanted to sync some ssh keys which have folder permission 0700, and file permission 0600. On one machine, I have permissions set properly. On another host that is sync, the permissions were all gone. Every folder is set to 0755, every file is to 0644. 

This means my ssh keys are world-readable on all machines except the original machine. That is not very secure. I have to manually change folder/file permissions on every machine.

**Comment by stuartm** [comment]
This issue really should get re-opened and addressed. The original conclusion that this is "not possible" is demonstrably false, there are ways to do it. If the actual reason for closing the ticket is "we're not willing to do it" then just say that, so we can all stop waiting for core problems like this to get resolved and move on.

**Comment by nipfuh** [comment]
I would like to see this implemented somehow as well, as something in my current setup regularly dies because of missing execute permissions.

**Comment by andresgongora** [comment]
I'm also in a spot where file permission sync-ing would make my life much easier. Yes, I see the problem when working cross-platform, but why not just safe the file permissions for each platform (nix/mac/windows) separately in the DB and then reapply them depending on what system the client is running on.

I use NC to sync lots of scripts between my machines and I have to constantly set the execution flag or, similar to @gengwg , manually fix the permissions for delicate stuff like keys that go along said scripts.

**Comment by traxeon** [comment]
If you can not get Nextcloud to do it, then take a look at unison file sync.   Consider the direction of the sync and use -auto -batch with a crontab entry.   I use it for correcting the .ssh key replication chmod that Nextcloud is not honoring.

**Comment by eliasrg** [comment]
For me, preserving permissions, or at least the executable bit, is also a basic re

**Comment by andresgongora** [comment]
Would it add additional latency or maintenance trouble if the file permissions were stored in the DB as @mimaoffice suggested?

**Comment by mochouinard** [comment]
Has this been fixed yet ?  It sadly one of the reason I still use Dropbox !

**Comment by marbx** [comment]
I have the same **feature request**.

**Comment by toptnc** [comment]
Please, re-open it.

**Comment by andresgongora** [comment]
This is just a *feature request*. Can't we leave it open at least? It's been 2 years since my last comment and I'd still find this very useful

**Comment by foerstermaniac** [comment]
I also stumbled upon this problem. I set up Nextcloud to sync my Workstations including Programming Projects and ssh-keys. Now i have to realize, that nextcloud is not suitable for this job, because of it's habit to change the permissions.

Please do as a favour and communicate your roadmap regarding this issue.

**Comment by C0rn3j** [comment]
@joshtrichards I see you tagged this issue, could you please also reopen it?

The reason why it was closed is invalid, as file permissions can be stored in the database, not just in the file system.

**Comment by jim-collier** [comment]
This isn't a "feature request", it's a bug. Preserving the exe bit is a basic function of file syncing.

I set up Nextcloud to replace Dropbox, but realized only after many many many hours of getting Nextcloud set up and running successfully, that I can't actually use Nextcloud.

I started using Dropbox nearly 15 years ago, in large part to synchronize settings and scripts across my multiple personal Linux instances and VMs. (Then later also for regular documents.) It wasn't perfect, but at least it maintained the execute bit - even when I accidentally synced linux and macos scripts to Windows, unnecessarily.

So for now, sadly I have to ditch Nextcloud after all that effort. It's such an impressive open-source accomplishment, and I wish I had the bandwidth to contribute to this issue directly.

**Comment by Rello** [comment]
Thanks a lot for your report. Much appreciated!

We understand your request, but we do not see a way to solve this in a reliable and stable way.

Nextcloud syncs file contents and metadata like names, and timestamps via webdav, but not file permissions like chmod. Permissions are considered system-specific metadata, which is managed locally by the operating system and not by Nextcloud.
File permission models vary between systems (e.g., Linux vs. Windows vs. Nextcloud), making it non-trivial for a sync client to handle them consistently.

There is also a security concern:
Syncing permissions like 0700 (restricting access to a specific user) could create security vulnerabilities if the file ends up on a less secure system.
Also private keys files must be protected. Syncing them through Nextcloud (or similar services) can expose them to risks unless the storage is secured (e.g., using server-side encryption).

We'd like to encourage to keep hacking. We're looking forward to your next report!

**Comment by C0rn3j** [comment]
@Rello I am glad Nextcloud considers file permissions on SSH keys important.

You might want to check out the ignored `2720677` report on HackerOne I sent you nearly 3 months ago.

**Comment by cpegel** [comment]
The reasons for closing this issue again after 7 years are still the same invalid ones and many in this thread have already suggested solutions:

There are no security concerns when storing the file permissions in the database instead of setting them at the server filesystem. Even more so: *not* syncing the permissions can lead to security concerns on the client side.

Different file permission models for different client operating systems are not a problem when you just store them separately for each OS.

Many in this thread also said that even if you don't want to go all the way and store all the permissions for all the operating systems, just the executable flag would still help a lot

I really hope you reconsider closing this for good and think about the valid use cases mentioned here over the years.

**Comment by mc6150** [comment]
Make it optional.
Warn of the 'risks'.
Make it a plug-in.

No shell script is executable because of the missing bit. It breaks the workflow.
Please reconsider.

**Comment by andresgongora** [comment]
I believe it would be beneficial to reconsider the decision to close this feature request. While storing and applying all possible file permissions and ownership options may not be entirely feasible, the ability to set specific permissions for secure files like SSH keys and remember execution flags for binaries could significantly enhance security and user experience.

For instance, on Windows, the Nextcloud Client could simply ignore unnecessary permission flags. This would allow for a more flexible and secure file management system within the Nextcloud environment.

Are there any technical limitations to this feature request that would making the implementation more difficult than it appears? It would only be an extra column in the database, and adding the logic to the Client to ensure that it stays consistent with the files on disk on the Client side; there is no need to apply any flags _inside_ the Server instance (for obvious reasons).

**Comment by jim-collier** [comment]
@Rello 

> we do not see a way to solve this in a reliable and stable way.

You mean, like pretty much all other open and closed-source sync products do - in a well-understood and verifiable way? (I.e. maintaining relevant metadata - eg exe bits - in own local database that gets synced.) You don't see a way to solve it, like that?

> There is also a security concern ... could create security vulnerabilities if the file ends up on a less secure system.

What contrived scenario are you envisioning here, exactly? It's 100% up to users, which of their own systems, their own files end up on. If they want their files on their own insecure system - keep in mind, a *syncronized* exe bit *that they themselves chose to apply*, is probably the least of their concerns, and not for anyone else to nanny for them. I'd rather a product or triage manager not artificially choose to cripple [or leave crippled] a potentially useful product based on nothing more than armchair speculation, with no real-world experience or evidence to back it up, out of a misguided desire to protect users from themselves.

But I guess it doesn't matter, for me anyway. At this point I have no choice but to bail. (And Thor knows I certainly don't have the time or bandwidth to learn the project and fix or fork it myself! Besides, it exists now in Seafile.) Good luck.

**Comment by C0rn3j** [comment]
Since the HackerOne report of mine is now closed and this issue is a wontfix, here's my report in full - 

The TL;DR is that while people here are mostly upset that permissions are taken away, it works both ways, and permissions you've *taken away* will be given back on a new sync too. 

![Image](https://github.com/user-attachments/assets/4928f7ad-4721-4e91-9b40-92cfdef66bf3)
